### PR TITLE
connectivity: add full egress gateway test suite

### PR DESCRIPTION
**Pull request by jibi**
see individual commits

Towards: https://github.com/cilium/cilium/issues/24151

### CODE REVIEW

**Comment by jibi** [comment]
test run: https://github.com/cilium/cilium/actions/runs/5112893978/jobs/9191464799?pr=25719

File: connectivity/check/context.go [hunk 1]
```diff
@@ -303,6 +303,12 @@ func (ct *ConnectivityTest) SetupAndValidate(ctx context.Context) error {
 	if err := ct.validateDeployment(ctx); err != nil {
 		return err
 	}
+	if ct.features.MatchRequirements(RequireFeatureEnabled(FeatureKPRExternalIPs)) &&
+		ct.features[FeatureFlavor].Mode != strings.ToLower(k8s.KindGKE.String()) {
```
Comment (brb) [inline connectivity/check/context.go:307]: Let's add a comment saying why we don't expose it on GKE.

File: connectivity/check/test.go [hunk 2]
```diff
@@ -445,13 +458,42 @@ func (t *Test) WithCiliumEgressGatewayPolicy(policy string) *Test {
...
+		// Set the excluded CIDRs
+		pl[i].Spec.ExcludedCIDRs = []v2.IPv4CIDR{}
+
+		switch params.ExcludedCIDRs {
+		case ExternalNodeExcludedCIDRs:
+			for _, nodeWithoutCilium := range t.NodesWithoutCilium() {
+				var node *corev1.Node
+
+				for _, client := range t.Context().Clients() {
+					n, err := client.GetNode(context.Background(), nodeWithoutCilium, metav1.GetOptions{})
```
Comment (brb) [inline connectivity/check/test.go:478]: You can use `NodesWithoutCiliumIPs` defined in `connectivity/check/check.go`.

File: connectivity/manifests/egress-gateway-policy-excluded-cidrs.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
+apiVersion: cilium.io/v2
+kind: CiliumEgressGatewayPolicy
+metadata:
+  name: cegp-sample-excluded-cidrs
+spec:
+  selectors:
+  - podSelector:
+      matchLabels:
+        io.kubernetes.pod.namespace: cilium-test
```
Comment (brb) [inline connectivity/manifests/egress-gateway-policy-excluded-cidrs.yaml:9]: Do we dynamically change it somewhere later on? Asking, as AFAIK `cilium-test` is configurable.

File: connectivity/tests/egressgateway.go [hunk 4]
```diff
@@ -39,151 +155,249 @@ func (s *egressGateway) Name() string {
...
+					DestCIDR:  "0.0.0.0/0",
+					EgressIP:  egressIP,
+					GatewayIP: egressGatewayNodeInternalIP.String(),
+				})
+		}
+
+		return targetEntries
+	})
+
+	// Ping hosts (pod to host connectivity)
```
Comment (brb) [inline connectivity/tests/egressgateway.go:199]: Maybe a small comment saying that this traffic should not be masqueraded.

File: connectivity/tests/egressgateway.go [hunk 5]
```diff
@@ -39,151 +155,249 @@ func (s *egressGateway) Name() string {
...
+			targetPodHostIP := echoPod.Pod.Status.HostIP
+			targetPodIP := echoPod.Pod.Status.PodIP
+
+			for _, externalNode := range ct.NodesWithoutCilium() {
+				for node, hostNetNSPod := range ct.HostNetNSPodsByNode() {
+					if node != externalNode {
+						continue
+					}
+
+					cmd := []string{"ip", "route", "add", targetPodIP, "via", targetPodHostIP}
```
Comment (brb) [inline connectivity/tests/egressgateway.go:284]: Could we use `WithIPRoutesFromOutsideToPodCIDRs()` instead? I know it's not fine-grained, but for your test case it should do the job.

File: connectivity/manifests/egress-gateway-policy-excluded-cidrs.yaml (hunk 3 above)
Comment (jibi) [reply connectivity/manifests/egress-gateway-policy-excluded-cidrs.yaml:9]: yup, it's already been taken care by WithCiliumEgressGatewayPolicy:

https://github.com/cilium/cilium-cli/blob/44ae1874fae4544c0db34dac89c11e37365b76ef/connectivity/check/test.go#L435-L446

File: connectivity/tests/egressgateway.go (hunk 5 above)
Comment (jibi) [reply connectivity/tests/egressgateway.go]: Nice, that works as well :rocket:

**Comment by jibi** [comment]
green run: https://github.com/cilium/cilium/actions/runs/5121292447/jobs/9208878978?pr=25719

File: connectivity/suite.go [hunk 6]
```diff
@@ -727,13 +732,25 @@ func Run(ctx context.Context, ct *check.ConnectivityTest, addExtraTests func(*ch
 		)
 
 	ct.NewTest("egress-gateway").
-		WithCiliumEgressGatewayPolicy(egressGatewayPolicyYAML).
+		WithCiliumEgressGatewayPolicy(egressGatewayPolicyYAML, check.CiliumEgressGatewayPolicyParams{}).
+		WithIPRoutesFromOutsideToPodCIDRs().
```
Comment (brb) [inline connectivity/suite.go:736]: Last nit (no need to re-run the CI): could you protect the test case with `if ct.Params().IncludeUnsafeTests {
` (as other `WithIPRoutesFromOutsideToPodCIDRs()` tests do)?

**Review by brb — APPROVED** [review]
:pizza:

**Review by tklauser — APPROVED** [review]

**Pull request merged**